File: TaxCalculator.java
// Problem Type: Array - 5
// LeetCode - 2303

// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : NO


// Your code here along with comments explaining your approach
/**
 * Algo:
 * 
 * a. Break the income into slabs such that pick the Math.min(remainingIncome, (currUpperBound - prevUpperBound)) as the current slab's taxable income
 */
/**
 * TC: If # of tax brackets are considered to be constant, then O(1)
 * else, O(n)
 */
/*
 * SC: O(1)
 */

public class TaxCalculator {
    
    public double calculateTax(int[][] brackets, int income) {
        int remIncome = income;
        double tax = 0;
        int prevBound = 0;
        int i = 0;
        int currIncome = 0;

        while(remIncome > 0) {
            currIncome = Math.min((brackets[i][0] - prevBound), remIncome);
            tax += currIncome * brackets[i][1] * 0.01;
            remIncome -= currIncome;
            prevBound = brackets[i][0];
            i++;
        }

        return tax;
    }
    
}

File: taxCalculator.cpp
// Time Complexity : O(N)
 // Space Complexity : O(1)
 // Did this code successfully run on Leetcode : Yes
 // Any problem you faced while coding this : NA


 // Your code here along with comments explaining your approach


double calculateTax(vector<vector<double>> levels, double salary) {
    double taxableSalary = 0.0;
    double tax = 0;
    double limit = 0.0;
    
    for(auto level : levels){
        double percent = level[1];
        if(level[0] == NULL){
            tax+=percent*salary;
            return tax;
        }
        taxableSalary = min(level[0]-limit,salary);
        tax+=taxableSalary*percent;
        limit = level[0];
        salary-=taxableSalary;
    }
    return tax;    
}
int main() {
    vector<vector<double>> levels;
    levels.push_back({10000.0, 0.1});
    levels.push_back({20000.0, 0.2});
    levels.push_back({30000.0, 0.3});
    levels.push_back({NULL, 0.4});
    cout<<calculateTax(levels, 25000);
    return 0;
}